# DATA LOADING


In [ ]:
import pandas as pd

In [ ]:
data = pd.read_csv('data_up.csv')

In [ ]:
data.head(5)

SEQN  Gender  Age  Annual-Family-Income  Ratio-Family-Income-Poverty  \
0     2       1   77                     8                         5.00   
1     5       1   49                    11                         5.00   
2    12       1   37                    11                         4.93   
3    13       1   70                     3                         1.07   
4    14       1   81                     5                         2.67   

   X60-sec-pulse  Systolic  Diastolic  Weight  Height  ...  Total-Cholesterol  \
0             68        98         56    75.4   174.0  ...               5.56   
1             66       122         83    92.5   178.3  ...               7.21   
2             64       174         99    99.2   180.0  ...               4.03   
3            102       130         66    63.6   157.7  ...               8.12   
4             72       136         61    75.5   166.2  ...               4.50   

    HDL  Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
0  1.39              4.7              3              3                 1   
1  1.08              5.5              1              1                 1   
2  0.98              5.2              2              1                 1   
3  1.28              7.6              3              3                 1   
4  1.04              5.8              1              1                 1   

   Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  CoronaryHeartDisease  
0         2                   2                 2                     0  
1         2                   2                 2                     0  
2         2                   1                 1                     0  
3         1                   1                 2                     0  
4         2                   2                 2                     0  

[5 rows x 51 columns]

In [ ]:
data.columns

Index(['SEQN', 'Gender', 'Age', 'Annual-Family-Income',
       'Ratio-Family-Income-Poverty', 'X60-sec-pulse', 'Systolic', 'Diastolic',
       'Weight', 'Height', 'Body-Mass-Index', 'White-Blood-Cells',
       'Lymphocyte', 'Monocyte', 'Eosinophils', 'Basophils', 'Red-Blood-Cells',
       'Hemoglobin', 'Mean-Cell-Vol', 'Mean-Cell-Hgb-Conc.',
       'Mean-cell-Hemoglobin', 'Platelet-count', 'Mean-Platelet-Vol',
       'Segmented-Neutrophils', 'Hematocrit', 'Red-Cell-Distribution-Width',
       'Albumin', 'ALP', 'AST', 'ALT', 'Cholesterol', 'Creatinine', 'Glucose',
       'GGT', 'Iron', 'LDH', 'Phosphorus', 'Bilirubin', 'Protein', 'Uric.Acid',
       'Triglycerides', 'Total-Cholesterol', 'HDL', 'Glycohemoglobin',
       'Vigorous-work', 'Moderate-work', 'Health-Insurance', 'Diabetes',
       'Blood-Rel-Diabetes', 'Blood-Rel-Stroke', 'CoronaryHeartDisease'],
      dtype='object')

In [ ]:
data.shape

(37079, 51)

In [ ]:
data.isnull().sum()

SEQN                           0
Gender                         0
Age                            0
Annual-Family-Income           0
Ratio-Family-Income-Poverty    0
X60-sec-pulse                  0
Systolic                       0
Diastolic                      0
Weight                         0
Height                         0
Body-Mass-Index                0
White-Blood-Cells              0
Lymphocyte                     0
Monocyte                       0
Eosinophils                    0
Basophils                      0
Red-Blood-Cells                0
Hemoglobin                     0
Mean-Cell-Vol                  0
Mean-Cell-Hgb-Conc.            0
Mean-cell-Hemoglobin           0
Platelet-count                 0
Mean-Platelet-Vol              0
Segmented-Neutrophils          0
Hematocrit                     0
Red-Cell-Distribution-Width    0
Albumin                        0
ALP                            0
AST                            0
ALT                            0
Cholesterol                    0
Creatinine                     0
Glucose                        0
GGT                            0
Iron                           0
LDH                            0
Phosphorus                     0
Bilirubin                      0
Protein                        0
Uric.Acid                      0
Triglycerides                  0
Total-Cholesterol              0
HDL                            0
Glycohemoglobin                0
Vigorous-work                  0
Moderate-work                  0
Health-Insurance               0
Diabetes                       0
Blood-Rel-Diabetes             0
Blood-Rel-Stroke               0
CoronaryHeartDisease           0
dtype: int64

Conclusion : Data is clean and all the data is numerical
CoronaryHeartDisease is target class. This is a classification problem predicting the heart disease.
SEQN is just representing the row number hence we need not take in the prediction columns.
Here, we have 49 features and 1 target column.
We have originally 37079 rows with no cell being empty.

# DATA ANALYSIS


In [ ]:
data.describe()

SEQN        Gender           Age  Annual-Family-Income  \
count  37079.000000  37079.000000  37079.000000          37079.000000   
mean   48901.041236      1.513282     48.943661              7.358208   
std    26753.636441      0.499830     18.010440              3.994083   
min        2.000000      1.000000     20.000000              1.000000   
25%    26120.500000      1.000000     33.000000              4.000000   
50%    50065.000000      2.000000     48.000000              7.000000   
75%    71173.500000      2.000000     63.000000             10.000000   
max    93702.000000      2.000000     85.000000             15.000000   

       Ratio-Family-Income-Poverty  X60-sec-pulse      Systolic     Diastolic  \
count                 37079.000000   37079.000000  37079.000000  37079.000000   
mean                      2.559026      72.579250    124.090078     69.919253   
std                       1.624789      12.242108     19.254741     13.575804   
min                       0.000000      32.000000      0.000000      0.000000   
25%                       1.140000      64.000000    111.000000     62.000000   
50%                       2.180000      72.000000    121.000000     70.000000   
75%                       4.130000      80.000000    134.000000     78.000000   
max                       5.000000     224.000000    270.000000    132.000000   

             Weight        Height  ...  Total-Cholesterol           HDL  \
count  37079.000000  37079.000000  ...       37079.000000  37079.000000   
mean      80.988276    167.389601  ...           5.081713      1.370344   
std       20.678734     10.122908  ...           1.072682      0.415985   
min       32.300000    129.700000  ...           1.530000      0.160000   
25%       66.500000    160.000000  ...           4.320000      1.070000   
50%       78.200000    167.100000  ...           5.020000      1.290000   
75%       92.100000    174.600000  ...           5.740000      1.600000   
max      371.000000    204.500000  ...          14.090000      5.840000   

       Glycohemoglobin  Vigorous-work  Moderate-work  Health-Insurance  \
count     37079.000000   37079.000000   37079.000000      37079.000000   
mean          5.676496       1.783840       1.598856          1.218587   
std           1.050223       0.448324       0.511199          0.461102   
min           2.000000       1.000000       1.000000          1.000000   
25%           5.200000       2.000000       1.000000          1.000000   
50%           5.400000       2.000000       2.000000          1.000000   
75%           5.800000       2.000000       2.000000          1.000000   
max          18.800000       3.000000       3.000000          9.000000   

           Diabetes  Blood-Rel-Diabetes  Blood-Rel-Stroke  \
count  37079.000000        37079.000000      37079.000000   
mean       1.907333            1.549502          1.796165   
std        0.349674            0.497550          0.402853   
min        1.000000            1.000000          1.000000   
25%        2.000000            1.000000          2.000000   
50%        2.000000            2.000000          2.000000   
75%        2.000000            2.000000          2.000000   
max        3.000000            2.000000          2.000000   

       CoronaryHeartDisease  
count          37079.000000  
mean               0.040670  
std                0.197527  
min                0.000000  
25%                0.000000  
50%                0.000000  
75%                0.000000  
max                1.000000  

[8 rows x 51 columns]

# Feature Selection


**ALGORITHM : BMPA**

In [ ]:
!ls -l


total 8184
-rw-r--r-- 1 root root 8374838 Oct 16 10:35 data_up.csv
drwxr-xr-x 1 root root    4096 Oct 14 13:41 sample_data


In [ ]:
!pip install mealpy


In [ ]:
import mealpy
import pkgutil
mods = [name for _, name, _ in pkgutil.walk_packages(mealpy.__path__, mealpy.__name__ + ".")]

print([m for m in mods if "mpa" in m.lower() or "bmpa" in m.lower()])


['mealpy.swarm_based.MPA']


In [ ]:
!pip install --upgrade git+https://github.com/thieu1995/mealpy.git scikit-learn pandas numpy -q

  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 91.2/91.2 kB 3.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.1/62.1 kB 2.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 9.5/9.5 MB 53.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.4/12.4 MB 39.4 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.3.3 which is incompatible.
dask-cudf-cu12 25.6.0 requires pandas<2.2.4dev0,>=2.0, but you have pandas 2.3.3 which is incompatible.
cudf-cu12 25.6.0 requires pandas<2.2.4dev0,>=2.0, but you have pandas 2.3.3 which is incompatible.


Imports


In [ ]:
import pandas as pd, numpy as np
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from mealpy.swarm_based.MPA import OriginalMPA
from mealpy.utils.space import BinaryVar

Load dataset


In [ ]:
df = pd.read_csv('data_up.csv')

In [ ]:
X = df.drop(columns=['SEQN', 'CoronaryHeartDisease'])

In [ ]:
y = df['CoronaryHeartDisease']

Here we are defining fitness function by (minimize 1- accuracy)
And also we Train/Test split



In [ ]:
def fitness(sol):
    mask = sol > 0.5
    if not mask.any():
        return 1
    Xs = X.loc[:, mask]

    Xtr, Xte, ytr, yte = train_test_split(Xs, y, test_size=0.2, random_state=42)
    clf = RandomForestClassifier(random_state=42).fit(Xtr, ytr)

    return 1 - accuracy_score(yte, clf.predict(Xte))

Binary Variables for MPA


In [ ]:
bounds = [BinaryVar() for _ in range(X.shape[1])]

Defining Optimization Problem




In [ ]:
problem = {
    "obj_func": fitness,
    "bounds": bounds,
    "minmax": "min",
}

Initializing Model and Running Binary MPA

In [ ]:
model = OriginalMPA(epoch=5, pop_size=20)
result = model.solve(problem)

INFO:mealpy.swarm_based.MPA.OriginalMPA:OriginalMPA(epoch=5, pop_size=20)
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 1, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 167.42458 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 2, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 175.34929 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 3, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 170.62871 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 4, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 151.28487 seconds
INFO:mealpy.swarm_based.MPA.OriginalMPA:>>>Problem: P, Epoch: 5, Current best: 0.04139697950377563, Global best: 0.04139697950377563, Runtime: 146.53248 seconds


Extracting best solution


In [ ]:
best_position = result.solution

if hasattr(result, "solution_fitness"):
    best_fitness = result.solution_fitness
elif hasattr(result, "fitness_history"):
    best_fitness = result.fitness_history[-1]
else:
    best_fitness = None
    print("Could not find fitness attribute in result. Inspect result.__dict__")



Could not find fitness attribute in result. Inspect result.__dict__


Selected Features


In [ ]:
selected_features = np.where(best_position > 0.05)[0]
X_selected = X.iloc[:, selected_features] if selected_features.size > 0 else pd.DataFrame()



Calculating the Actual Accuracy and printing the results

In [ ]:
if X_selected.shape[1] > 0:
    clf_final = RandomForestClassifier(random_state=42)
    accuracy = np.mean(cross_val_score(clf_final, X_selected, y, cv=5))
else:
    accuracy = 0



In [ ]:
print("Best Fitness (1 - Accuracy):", best_fitness)


Best Fitness (1 - Accuracy): None


In [ ]:
print("Selected Features:", selected_features)


Selected Features: [ 0  1  2  3  4  6  7  9 10 11 13 14 15 16 17 19 20 22 24 27 28 30 31 32
 34 35 36 37 39 40 41 43 44 45 47 48]


In [ ]:
print("Total Features Selected:", len(selected_features))


Total Features Selected: 36


In [ ]:
print("Accuracy of Selected Features:", accuracy)

Accuracy of Selected Features: 0.9593570498006532


**ALGORITHM : EBPSO**

Installing MealPy

In [ ]:
!pip install mealpy --quiet


Importing Libraries

In [ ]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

from mealpy.swarm_based.PSO import C_PSO
from mealpy.utils.space import BinaryVar


Loading dataset

In [ ]:
df = pd.read_csv('data_up.csv')
X, y = df.iloc[:, :-1], df.iloc[:, -1]



Ensuring X is DataFrame and y is ID array

In [ ]:

X = pd.DataFrame(X)
y = np.array(y).ravel()

Defining Fitness Function and Train/Test split

In [ ]:
def fitness(sol):
    mask = sol > 0.5
    if not mask.any():
        return 1
    Xs = X.loc[:, mask]

    Xtr, Xte, ytr, yte = train_test_split(Xs, y, test_size=0.2, random_state=42)
    clf = RandomForestClassifier(random_state=42).fit(Xtr, ytr)

    return 1 - accuracy_score(yte, clf.predict(Xte))


Defining Binary Variables

In [ ]:
bounds = [BinaryVar() for _ in range(X.shape[1])]
problem = {
    "obj_func": fitness,
    "bounds": bounds,
    "minmax": "min",
}


Run

In [ ]:
model = C_PSO(epoch=3, pop_size=20)
result = model.solve(problem)


INFO:mealpy.swarm_based.PSO.C_PSO:C_PSO(epoch=3, pop_size=20, c1=2.05, c2=2.05, w_min=0.4, w_max=0.9)
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 1, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 314.74196 seconds
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 2, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 306.38196 seconds
INFO:mealpy.swarm_based.PSO.C_PSO:>>>Problem: P, Epoch: 3, Current best: 0.04153182308522119, Global best: 0.04153182308522119, Runtime: 308.53841 seconds


Extracting Best Solution

In [ ]:
best_position = result.solution
best_fitness = getattr(result, "solution_fitness", None) or getattr(result, "fitness_history", [None])[-1]


Calculating Final Accuracy

In [ ]:
if X_selected.shape[1] > 0:
    clf_final = RandomForestClassifier(random_state=42)
    accuracy = np.mean(cross_val_score(clf_final, X_selected, y, cv=5))
else:
    accuracy = 0


Printing Results

In [ ]:
print("Best Fitness (1 - Accuracy):", best_fitness)
print("Selected Features Indices:", selected_features)
print("Total Features Selected:", len(selected_features))
print("Accuracy of Selected Features:", accuracy)
